libcontainer/intelrdt/intelrdt.go

```diff
@@ -0,0 +1,556 @@
+		return nil, err
+	}
+	defer f.Close()
+
+	var (
+		s   = bufio.NewScanner(f)
+		out = []int{}
+	)
+
+	for s.Scan() {
+		if t := s.Text(); t != "" {
+			pid, err := strconv.Atoi(t)
+			if err != nil {
+				return nil, err
+			}
+			out = append(out, pid)
+		}
+	}
+	return out, nil
+}
+
+func writeFile(dir, file, data string) error {
+	if dir == "" {
+		return fmt.Errorf("no such directory for %s", file)
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, file), []byte(data+"\n"), 0700); err != nil {
+		return fmt.Errorf("failed to write %v to %v: %v", data, file, err)
+	}
+	return nil
+}
+
+func getIntelRdtData(c *configs.Config, pid int) (*intelRdtData, error) {
+	rootPath, err := getIntelRdtRoot()
+	if err != nil {
+		return nil, err
+	}
+	return &intelRdtData{
+		root:   rootPath,
+		config: c,
+		pid:    pid,
+	}, nil
+}
+
+// WriteIntelRdtTasks writes the specified pid into the "tasks" file
+func WriteIntelRdtTasks(dir string, pid int) error {
+	if dir == "" {
+		return fmt.Errorf("no such directory for %s", IntelRdtTasks)
+	}
+
+	// Dont attach any pid if -1 is specified as a pid
+	if pid != -1 {
+		if err := ioutil.WriteFile(filepath.Join(dir, IntelRdtTasks), []byte(strconv.Itoa(pid)), 0700); err != nil {
+			return fmt.Errorf("failed to write %v to %v: %v", pid, IntelRdtTasks, err)
+		}
+	}
+	return nil
+}
+
+// Check if Intel RDT is enabled
+func IsIntelRdtEnabled() bool {
+	// 1. check if hardware and kernel support Intel RDT feature
+	// "rdt" flag is set if supported
+	isFlagSet, err := parseCpuInfoFile("/proc/cpuinfo")
+	if err != nil {
+		return false
+	}
+
+	// 2. check if Intel RDT "resource control" filesystem is mounted
+	isMounted := isIntelRdtMounted()
+
+	return isFlagSet && isMounted
+}
+
+// Get Intel RDT "resource control" filesystem path
+func GetIntelRdtPath(id string) (string, error) {
+	rootPath, err := getIntelRdtRoot()
+	if err != nil {
+		return "", err
+	}
+
+	path := filepath.Join(rootPath, id)
+	return path, nil
+}
+
+// Get read-only Intel RDT related system information
+func GetIntelRdtInfo() (*IntelRdtInfo, error) {
+	intelRdtInfo := &IntelRdtInfo{}
+
+	rootPath, err := getIntelRdtRoot()
+	if err != nil {
+		return nil, err
+	}
+
+	path := filepath.Join(rootPath, "info", "l3")
+	cbmMask, err := getIntelRdtParamUint(path, "cbm_mask")
+	if err != nil {
+		return nil, err
+	}
+	minCbmBits, err := getIntelRdtParamUint(path, "min_cbm_bits")
+	if err != nil {
+		return nil, err
+	}
+	numClosid, err := getIntelRdtParamUint(path, "num_closid")
+	if err != nil {
+		return nil, err
+	}
+
+	intelRdtInfo.CbmMask = cbmMask
+	intelRdtInfo.MinCbmBits = minCbmBits
+	intelRdtInfo.NumClosid = numClosid
+
+	return intelRdtInfo, nil
+}
+
+// Applies configuration to the process with the specified pid
+func (m *IntelRdtManager) Apply(pid int) (err error) {
+	d, err := getIntelRdtData(m.Config, pid)
+	if err != nil && !IsNotFound(err) {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	path, err := d.join(m.Id)
```

Comment:
@yummypeng 

For the nested containers case, I added comments in:
https://github.com/opencontainers/runc/pull/1279#discussion_r114473075